### PR TITLE
Docs model

### PR DESCRIPTION
**Pull request by ajdapretnar**
##### Issue
Documentation was obsolete since merging Classify and Regression categories (panes).

##### Description of changes
- Classify and Regression merged into Model category.
- Images and icons updated for Model category.
- Links to former categories updated.
- Obsolete images elsewhere partially updated.

##### Includes
- [X] Documentation

### CODE REVIEW

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+The **Stochastic Gradient Descent** widget uses `stochastic gradient descent <https://en.wikipedia.org/wiki/Stochastic_gradient_descent>`_ that minimizes a chosen loss function with a linear function. The algorithm approximates a true gradient by considering one sample at a time, and simultaneously updates the model based on the gradient of the loss function. For regression, it returns predictors as minimizers of the sum, i.e. M-estimators, and is especially useful for large-scale and sparse data sets.
+
+.. figure:: images/StochasticGradientDescent-stamped.png
+   :scale: 50 %
+
+1. Specify the name of the model. The default name is "SGD".
+2. Algorithm parameters. Classification loss function:
+
+   -  `Hinge <>`_ ()
```
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: @pavlin-policar Could you please check Stochastic Gradient (and SVM, too), since you were our chief 'merger'. These are the new additions that were not found in the old documentation. You can either add to this PR or just write the missing things in the comment.

File: doc/visual-programming/source/widgets/model/svm.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+`Support vector machine <https://en.wikipedia.org/wiki/Support_vector_machine>`_ (SVM) is a machine learning technique that separates the attribute space with a hyperplane, thus maximizing the margin between the instances of different classes or class values. The technique often yields supreme predictive performance results. Orange embeds a popular implementation of SVM from the `LIBSVM <https://www.csie.ntu.edu.tw/~cjlin/libsvm/>`_ package. This widget is its graphical user interface.
+
+For regression tasks, **SVM** performs linear regression in a high dimension feature space using an ε-intensive loss. Its estimation accuracy depends on a good setting of C, ε and kernel parameters. The widget outputs class predictions based on a `SVM Regression <https://en.wikipedia.org/wiki/Support_vector_machine#Regression>`_.
+
+The widget works for both classification and regression tasks.
+
+.. figure:: images/SVM-stamped.png
+   :scale: 50 %
+
+1. The learner can be given a name under which it will appear in other widgets. The default name is "SVM".
```
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/svm.rst]: @pavlin-policar Check this too, please. :)

**Comment by ajdapretnar** [comment]
I don't understand why the build fails. I built the files locally and there were no errors.
Also, the file that Travis is referring to is deleted, so what the heck?

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+-  **Preprocessor**
+
+   Preprocessing method(s)
+
+**Outputs**:
+
+-  **Learner**
+
+   A SGD regression learning algorithm with settings as specified in the dialog.
```
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: It shouldn't say `regression` here.

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 1 above)
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: A hinge loss is used by linear SVMs, so we could put `(linear SVM)`.

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+   :scale: 50 %
+
+1. Specify the name of the model. The default name is "SGD".
+2. Algorithm parameters. Classification loss function:
+
+   -  `Hinge <>`_ ()
+   -  `Logistic Regression <>`_ ()
+   -  `Modified Huber <>`_ ()
+   -  `Squared Hinge <>`_ ()
+   -  `Perceptron <>`_ ()
```
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: This is another linear loss used by the perceptron algorithm. I'm really just getting all this from the [scikit-learn docs](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier).

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      (mixing both penalty norms).
+
+   Regularization strength defines how... and mixing parameter how ....
+
+4. Learning parameters.
+
+   - Learning rate:
+
+      - *Constant*: learning rate stays the same through all epochs (passes)
+      - *Optimal*:
```
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: The optimal learning rate is chosen by a heuristic proposed by Leon Bottou.

File: doc/visual-programming/source/widgets/model/svm.rst (hunk 2 above)
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/svm.rst]: The description seems fine, but I would like to point out something that could potentially be confusing.

With the regular SVM, there are two settings
1. **C** is just the penalty term for the loss and applies everywhere.
2. **epsilon** is a parameter to the epsilon-SVR model, and only applies to regression tasks (unfortunately, since the learner can't know what type of data it will receive in advance, it has to be enabled all the time. I've tried to make it clear that it will only be applied to regression tasks with *Regression* in the name.

The nu-SVM is similar:
1. **C** only applies to regression tasks here and is the penalty of the error.
2. **nu** applies to both problems.

**Comment by pavlin-policar** [comment]
I haven't figured out how to add commits to the pull request, so I've left some comments. Both the SGD and SVM widgets are essentially wrappers to scikit-learn learners, so their docs could come in handy.

Also, when building the docs with `make html`, I also get build errors, but the build does apparently finish. Hopefully this helps.

```
orange3/doc/visual-programming/source/widgets/model/stochasticgradient.rst:42: ERROR: Unknown target name: "hinge <>".
orange3/doc/visual-programming/source/widgets/model/stochasticgradient.rst:43: ERROR: Unknown target name: "logistic regression <>".
orange3/doc/visual-programming/source/widgets/model/stochasticgradient.rst:44: ERROR: Unknown target name: "modified huber <>".
orange3/doc/visual-programming/source/widgets/model/stochasticgradient.rst:45: ERROR: Unknown target name: "squared hinge <>".
orange3/doc/visual-programming/source/widgets/model/stochasticgradient.rst:46: ERROR: Unknown target name: "perceptron <>".
orange3/doc/visual-programming/source/widgets/model/svm.rst:75: WARNING: image file not readable: widgets/model/images/SVM-predictions.png
```

**Comment by ajdapretnar** [comment]
Thanks!

These build errors are just to remind me/us what is missing in docs. This needs to be completed before the documentation is submitted. :) (apart from the last one, which I didn't get on my local build 😮 )

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 3 above)
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: Ups. Fixed.

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 1 above)
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: You mean SGD?

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 4 above)
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: I entered what I could from sklearn, but not all options are covered or explained. I'll need some help with those.

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+   -  None.
+   -  `Lasso (L1) <https://en.wikipedia.org/wiki/Taxicab_geometry>`_ (L1,
+      leading to sparse solutions)
+   -  `Ridge (L2) <https://en.wikipedia.org/wiki/Norm_(mathematics)#p-norm>`_
+      (L2, standard regularizer)
+   -  `Elastic net <https://en.wikipedia.org/wiki/Elastic_net_regularization>`_
+      (mixing both penalty norms).
+
+   Regularization strength defines how... and mixing parameter how ....
```
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: @pavlin-policar Need help. I don't know what they do and Google is not helpful here (for me).

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+4. Learning parameters.
+
+   - Learning rate:
+
+      - *Constant*: learning rate stays the same through all epochs (passes)
+      - *Optimal*: 
+      - `Inverse scaling <http://users.ics.aalto.fi/jhollmen/dippa/node22.html>`_: earning rate is inversely related to the number of iterations
+
+   - Initial learning rate:
+   - Inverse scaling exponent:
```
Comment (ajdapretnar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: @pavlin-policar The above two as well. Are they self-explanatory or would the user need some tips on what they are for?

File: doc/visual-programming/source/widgets/model/svm.rst [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+-  **Support Vectors**
+
+   A subset of data instances from the training set that were used as support vectors in the trained model.
+
+Description
+-----------
+
+`Support vector machine <https://en.wikipedia.org/wiki/Support_vector_machine>`_ (SVM) is a machine learning technique that separates the attribute space with a hyperplane, thus maximizing the margin between the instances of different classes or class values. The technique often yields supreme predictive performance results. Orange embeds a popular implementation of SVM from the `LIBSVM <https://www.csie.ntu.edu.tw/~cjlin/libsvm/>`_ package. This widget is its graphical user interface.
+
+For regression tasks, **SVM** performs linear regression in a high dimension feature space using an ε-intensive loss. Its estimation accuracy depends on a good setting of C, ε and kernel parameters. The widget outputs class predictions based on a `SVM Regression <https://en.wikipedia.org/wiki/Support_vector_machine#Regression>`_.
```
Comment (kernc) [inline doc/visual-programming/source/widgets/model/svm.rst]: ε-insensitive

**Comment by kernc** [comment]
> I don't understand why the build fails. The file that Travis is referring to is deleted, so what the heck?

I don't fully understand either. I assume `git diff` was returning that filename for some reason. I pushed https://github.com/biolab/orange3/pull/2193/commits/7657b8dcc84802b676c47ea020a7882ce128b37a which makes it a bit more robust in this event. Now the build should still fail, but the reason is clearer.

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 1 above)
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: No, SGD is just the optimization method. If we optimize a hinge loss with SGD, we get a linear SVM, we've just gotten to it using a different method (I'm guessing the regular SVM in sklearn uses lbfgs or some other super optimized method). Similarly, if we optimize the log loss, we get logistic regression or squared loss for regression, we get linear regression. This is only ever really useful when we have a lot of data.

Imo, it's kind of a strange separation on scikit-learn's part... It's also strange they don't implement mini-batch gradient descent if they've already gone with sgd...

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 4 above)
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: So, logistic regression and perceptron should be quite self-explanatory, hinge loss is just a linear SVM, `modified_huber` is apparently a smoothed version of a linear SVM. `squared_hinge` from what I can tell just penalized outliers more (*like hinge but is quadratically penalized*). These are shown quite nicely by [this](https://en.wikipedia.org/wiki/Hinge_loss#/media/File:Hinge_loss_variants.svg) figure. I'm not really sure how why one would be better than the other though...

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 6 above)
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: My bad, I totally missed this :) The regularization strength should be pretty self explanatory: the lower the value, the less we regularize, the more we allow the model to exactly fit the data.

Elastic net is basically a mix between the L1 and L2 loss, and the mixing parameter tells us what their ratio should be. The formula is something like this: mix * L1 + (1-mix) * L2.

So if the mixing parameter is set to 0, we get the L2 loss, and if we set it to 1, we get the L1 loss.

File: doc/visual-programming/source/widgets/model/stochasticgradient.rst (hunk 7 above)
Comment (pavlin-policar) [inline doc/visual-programming/source/widgets/model/stochasticgradient.rst]: The initial learning rate should be fairly self explanatory. The inverse scaling factor seems to be some kind of learning rate decay, which just means that at first we want to make larger steps towards the minimum of the loss function, but later, we expect to be closer to the minimum so we want to make smaller steps.

I'm not really sure about this, because the scikit learn wiki is useless here but by looking at the formula, I don't really know what else it could be.

**Comment by ajdapretnar** [comment]
> I don't fully understand either. I assume git diff was returning that filename for some reason. I pushed 7657b8d which makes it a bit more robust in this event. Now the build should still fail, but the reason is clearer. Make sure you fetch and rebase (or pull) first before pushing.

Sorry, sorry, sorry! I didn't read this thoroughly! 😢 I already rebased, is there any change of getting the commit back? 🙈

**Comment by kernc** [comment]
The commit is still there somewhere since GH is showing it. I don't know where to fetch it from to check it out. You can always apply the patch:
```bash
$ wget -O commit.patch https://github.com/biolab/orange3/commit/7657b8dcc84802b676c47ea020a7882ce128b37a.patch
$ cat commit.patch
$ git apply commit.patch
```

**Comment by ajdapretnar** [comment]
Do I have to add the commit.patch as well, or just the modified file?

**Comment by kernc** [comment]
It's not `git apply`, it's **`git am`** to apply with authorship information. :+1:

**Comment by kernc** [comment]
Right, so the build fails because [not all modified images have indexed palettes](https://travis-ci.org/biolab/orange3/jobs/224258392#L2287).
https://docs.gimp.org/en/gimp-image-convert-indexed.html

**Comment by ajdapretnar** [comment]
Yes, they are icons. Do they really have to be indexed? This is wasting a lot of my time on indexing.... 😞

**Comment by kernc** [comment]
Yes, they do. For PNG images this can be easily automated:
```bash
$ convert -format PNG8 input.png output.png  # Will create output.png with indexed palette
# or ...
$ mogrify -format PNG8 *.png  # Will index all images in current directory **in-place**

**Comment by astaric** [comment]
@ajdapretnar @kernc is this ready to be merged?

**Comment by ajdapretnar** [comment]
I think so.
